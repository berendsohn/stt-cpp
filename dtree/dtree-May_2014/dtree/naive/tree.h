// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_NAIVE_TREE_H_
#define DTREE_NAIVE_TREE_H_

#include "dtree/naive/common.h"

namespace naive {

template<typename Base>
class EndTree : public Base {
 public:
  EndTree() : parent_(NULL) {}
  EndTree* parent() const { return parent_; }
  void set_parent(EndTree* parent) { parent_ = parent; }
  template<typename Forwarder>
  void CopyFrom(const EndTree& node, Forwarder forward);
 private:
  EndTree* parent_;
  DISALLOW_COPY_AND_ASSIGN(EndTree);
};

template<typename Type, typename Base>
class WithStaticValue : public Base {
 public:
  typedef dtree::Nop<Type> Group_;
  Type value() const { return value_; }
  void set_value(Type value) { value_ = value; }
  //
  template<typename Node>
  static Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, Type value);
 protected:
  WithStaticValue() : value_() {}
 private:
  Type value_;
};

template<typename Group, typename Base>
class WithAncValue : public Base {
 public:
  typedef Group Group_;
  typename Group::Type value() const { return value_; }
  void set_value(typename Group::Type value) { value_ = value; }
  void add_to_value(typename Group::Type delta) {
    set_value(Group::Plus(value(), delta));
  }
  void subtract_from_value(typename Group::Type delta) {
    set_value(Group::Minus(value(), delta));
  }
  //
  template<typename Node>
  static typename Group::Type Value(Node* node);
  template<typename Node>
  static void SetValue(Node* node, typename Group::Type value);
  template<typename Node>
  static void AddToAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void AddToProperAnc(Node* node, typename Group::Type delta);
  template<typename Node>
  static void SubtractFromProperAnc(Node* node, typename Group::Type delta);

 protected:
  WithAncValue() : value_() {}

 private:
  typename Group::Type value_;
};

template<typename Semigroup, typename Base>
class WithAncAggr : public Base {
 public:
  typedef Semigroup Semigroup_;
  typename Semigroup::Type singleton_aggr() const {
    return Semigroup::AggrFromValue(this->value());
  }
  //
  template<typename Node>
  static typename Semigroup::Type AggrAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostAnc(int dir, Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostAnc(Node* node, const Predicate& predicate);
  template<typename Node>
  static typename Semigroup::Type AggrProperAnc(Node* node);
  template<typename Node, typename Predicate>
  static Node* FindDirmostProperAnc(int dir,
                                    Node* node,
                                    const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindLeafmostProperAnc(Node* node, const Predicate& predicate);
  template<typename Node, typename Predicate>
  static Node* FindRootmostProperAnc(Node* node, const Predicate& predicate);

 protected:
  WithAncAggr() {}
};

template<typename Base>
class WithEvertBy : public Base {
 private:
  typedef typename Base::Group_ Group;
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvertBy() {}
};

template<typename Base>
class WithEvert : public Base {
 public:
  template<typename Node>
  static void Evert(Node* node);
 protected:
  WithEvert() {}
};
}  // namespace naive
#endif  // DTREE_NAIVE_TREE_H_
