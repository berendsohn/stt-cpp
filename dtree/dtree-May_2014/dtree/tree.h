// -*-C++-*-
// Copyright (c) 2012--2014 David Eisenstat <eisenstatdavid@gmail.com>
//     and Brown University
// Released under http://opensource.org/licenses/MIT
// May 2014 version

#ifndef DTREE_TREE_H_
#define DTREE_TREE_H_

#include "dtree/selector.h"

#ifdef NAIVE_DTREE
#include "dtree/naive/tree.h"
#else
#include "dtree/selfadjust/tree.h"
#endif
#endif  // DTREE_TREE_H_
