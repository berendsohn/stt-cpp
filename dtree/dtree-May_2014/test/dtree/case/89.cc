#define DTREE_CONFIG EndTreeWithDesc<WithDescAggrOfDescValue<dtree::DpMin<double>, dtree::DwLessEqual_<double>, WithStaticValue<dtree::DpValue<double>, WithAncDescAggr<dtree::Count<int>, NullPredicate_, Begin > > > >
#include "test/dtree/tree.cc"
