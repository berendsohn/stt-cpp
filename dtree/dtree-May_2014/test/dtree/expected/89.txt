<<<
  1: vL4R3F0
  2: vL1R4F1
  3: vL1R-1F1
  4: vL-2R0F1 -> 10
  5: vL4R-1F0
  6: vL2R2F0 -> 4
  7: vL-6R-8F1
  8: vL0R-6F1 -> 14
  9: vL5R1F1
  10: vL-7R-6F0 -> 8
  11: vL-5R-1F1 -> 12
  12: vL6R-6F0 -> 6
  13: vL1R-3F1 -> 8
  14: vL6R2F1
  15: vL-4R7F1 -> 8
>>>
<<<
  @1: a1 o0 vL4R3F0 a1 o0 vL4R3F0 aL4R3

  @2: a2 o0 vL1R4F1 a2 o0 vL1R4F1 aL-5R-1

  @3: a1 o0 vL1R-1F1 a1 o0 vL1R-1F1 aL1R-1

  @5: a1 o0 vL4R-1F0 a1 o0 vL4R-1F0 aL4R-1

  @7: a1 o2 vL-6R-8F1 a1 o2 vL-6R-8F1 aL-6R-8

  @9: a1 o0 vL5R1F1 a1 o0 vL5R1F1 aL5R1

  @14: a1 o0 vL6R2F1 a1 o0 vL6R2F1 aL6R2
      /8: a3 o2 vL0R-6F1 a3 o2 vL0R-6F1 aL-7R-6
         .'13: a1 o1 vL1R-3F1 a1 o1 vL1R-3F1 aL1R-3
       ->15: a1 o2 vL-4R7F1 a1 o2 vL-4R7F1 aL-4R-3
          /10: a1 o0 vL-7R-6F0 a1 o0 vL-7R-6F0 aL-7R-6
        \4: a2 o0 vL-2R0F1 a2 o0 vL-2R0F1 aL-7R-6
    \6: a6 o2 vL2R2F0 a6 o2 vL2R2F0 aL-7R-6
      \12: a2 o0 vL6R-6F0 a2 o0 vL6R-6F0 aL-5R-6
        \11: a1 o0 vL-5R-1F1 a1 o0 vL-5R-1F1 aL-5R-1
>>>
